### ISSUE TITLE
Reconsidering signature of `kn service create`

### ISSUE THREAD

**Issue by rhuss** [knative/client#166]
At the moment, `kn service` create takes a mandatory argument (the service name) and a mandatory option `--image` for the image to run.

Having a mandatory _option_ (an oxymoron) is kind of an anti-pattern, and I think we should avoid that. Also, a name doesn't have to be mandatory as a default name can be extracted from the image name.

My suggestion would be that instead of

```
./kn service create hello --image gcr.io/knative-samples/helloworld-go
```

we use

```
./kn service create gcr.io/knative-samples/helloworld-go
```

with an (optional :) option `--name` to specify the name, e.g.

```
./kn service create gcr.io/knative-samples/helloworld-go --name hello

```

This leads to a more concise and also shorter usage pattern. A minor drawback would be that `kn service update` would not be symmetrical to `kn service create` anymore (update would need the name as mandatory argument, and the image as option), but I believe that is ok.

**Comment by duglin** [comment]
I'm a huge +10000 on converting `--image` into a required position parameter. I agree with you that "mandatory options" are weird and that fact that `kubectl` does it doesn't mean they made the right choice :-)    I even tried to find a standard unix/linux command that follows that pattern to see why they did it but have yet to find one.

As for defaulting the name, I'm not as keen on that one since I'm not sure that a majority of the time people would be happy with the image name as the KnService name.  But TBH I have no data to back it up, just a gut feel.  And I'd be happy to ignore that feeling if we can get rid of "mandatory options"  LOL

**Comment by rhuss** [comment]
It's a good question, whether having a name automatically inferred by default is a good choice or not.

My argumentation is that even worse than having mandatory options would be to have more than one mandatory argument because then you have to memorize the order of those. In that case (i.e. both name and image are mandatory) I would even opt to make both options, and don't call it options but 'named arguments' ;-) (and they can be given in any order)

We already have the first two arguments slots reserved for command and subcommand, and then maybe one argument (that already make 4 parts including 'kn'). But not two or more arguments to the subcommand, please.

My preference would be

* If name can be made optional (i think so):
   - ./kn service create hello gcr.io/knative-samples/helloworld-go
   - ./kn service create hello gcr.io/knative-samples/helloworld-go --name hello
* If name is mandatory:
   - ./kn service create --name hello --image gcr.io/knative-samples/helloworld-go
   - ./kn service create --image gcr.io/knative-samples/helloworld-go --name hello

**Comment by duglin** [comment]
I've often seen linux commands with optional args w/o flags, so it would look like this:
```
$ kn service create IMAGE [ NAME ]
$ kn service create duglin/echo
$ kn service create duglin/echo echo    # same a previous
$ kn service create duglin/echo repeat  # new name
```
This to me is similar to how cmds like `grep` are of the form:
```
$ grep [OPTIONS] PATTERN [FILE...]
```
where I'm equating FILE to our NAME. This, kind of, gives the best of both - optional but when specified doesn't need a pesky `--name`

**Comment by rhuss** [comment]
Bu this works also only when name is optional, right ?

> grep [OPTIONS] PATTERN [FILE...]

has also only _one_ mandatory argument, and grep imo uses not an option but arguments for the filenames because it a vararg and hence wildcards can be used. You can use `grep sth *` which doesn't work with `grep sth --file *`.  That's not the case in our example as we deal with a single name only.

**Comment by duglin** [comment]
I'm suggesting that we do what you want... make name optional. But we can do that my moving it to the end.  To me it's the same as grep because to me the reason for the optional positional arg (FILE) doesn't matter to me (I'm ignoring the fact FILE can be repeated since that's not relevant). When the arg is present we use it. When it is absent we default. And in grep's case that default is "use stdin" and in our case its "use the image name"

**Comment by duglin** [comment]
If you'd like, consider `ln [OPTION]... [-T] TARGET LINK_NAME` which has more than one required positional args. Both required in this case though. Or `uniq [OPTION]... [INPUT [OUTPUT]]`. The pattern I see is that optional positional args are common as long as they're at the end (after the required positional args) but the ordering of those optional ones does matter when there's more than one type of arg (not a repeated one), as in the case of `uniq`

**Comment by sixolet** [comment]
The reason I made it a flag initially was for consistency with the update command, where it is and should remain a flag.

**Comment by rhuss** [comment]
I see that pattern. As mentioned in the description: It depends whether we value symmetry between create & update over conciseness and semantic consistency (i.e. options being optional).

```
kn service create gcr.io/knative-samples/helloworld-go
kn service update helloworld-go --env KEY=VALUE
```

vs

```
kn service create hello-world --image gcr.io/knative-samples/helloworld-go
kn service update hello-world --env KEY=VALUE
```

vs

```
kn service create --name hello-world --image gcr.io/knative-samples/helloworld-go
kn service update --name hello-world --env KEY=VALUE
```

**Comment by rhuss** [comment]
@duglin `ln` has only 1 mandatory argument, `uniq` has no mandatory argument. In fact, there are very rare cases where you find more than one mandatory argument.

**Comment by rhuss** [comment]
I didn't want to open Pandora's box, I think there are arguments for all variations. If there is no common agreement, I will be fine with like it is. The only thing which probably goes beyond the matter of taste is whether we would like to make the name optional for create, defaulting it to the image name (but keep it of course mandatory for update).

**Comment by duglin** [comment]
> The reason I made it a flag initially was for consistency with the update command, where it is and should remain a flag.

> It depends whether we value symmetry between create & update over conciseness and semantic consistency

Consistency is great! But only when it makes sense. If the same value CAN appear on two different commands, but in one case it's OPTIONAL and in the other it's REQUIRED, then I see no reason why we need to be consistent w.r.t. the present of a `--flag` since the use of a flag is usually about its optionality or not.  Or put another way, on 'update' the image name is just one of many values that can be optionally passed in so a flag on that is basically required to disambiguate it from other options.

I really think: `kn service create [ flags ] IMAGE [ NAME ]` gives you exactly what you're looking for. Does anyone object to this?

**Comment by maximilien** [comment]
I actually like @duglin last suggestion. So the `NAME` parameter is optional and positional after the required `IMAGE` and as was initially suggested if the name is not present it generated with the image name and some hash. So something like:

```bash
kn service create gcr.io/knative/hello-world —env TARGET=Defaulted
```

Creates a service with name like: `hello-world-1234`. And the following:

```bash
kn service create gcr.io/knative/hello-world hello —env TARGET=Named
```

Creates a service with name like: `hello`

——
A nice side effect — might have been mentioned before — is the ability to create services that you know are guaranteed to have a unique name. This becomes an issue when you have many instances of a service. Though I am not sure how common this use case (multiple instance of a service) is outside of demos.

**Comment by rhuss** [comment]
I don't like that hash part for everything which is created manually. It totally makes your bash history unusable, forces you more to memorize the name and cause you to call `kn service get` just to find out the name again.

And how often do you create two services with the same image  ? I would just error out if the service already exists and the user has not provided `--force`.

**Comment by duglin** [comment]
If we don't require a name I think we need to have a well defined, and child mind calculable, pattern for what the name will be. Using the image name for now seems ok, and then erroring if it's already in use, like we do today.

**Comment by sixolet** [comment]
I must confess I'm a little dubious about automatically naming your service, since it's the thing you always have to name to refer to later for any other operation.

**Comment by rhuss** [comment]
Isn't it that you can still super easily name the service as you want? It's just that you don't _have_ to as we have a reasonably good default value (convention over configuration). If people stick to this default convention they would also have automatically a direct correlation back to (part of) the image name, which might be helpful if you see a service's name without having access to the internals.

**Comment by maximilien** [comment]
OK so auto-name is out. So looks like we have two possible approached in contention:

1. sequence name from the image name

```bash
kn service create gcr.io/knative/hello-world —env TARGET=Defaulted
Service 'hello-world' created in 'default' namespace.

kn service create gcr.io/knative/hello-world —env TARGET=Defaulted
Service 'hello-world-1' created in 'default' namespace.

# and so on
```

2. fail unless `--force`

```bash
kn service create gcr.io/knative/hello-world hello-world —env TARGET=Defaulted
Service 'hello-world' created in 'default' namespace.

...

kn service create gcr.io/knative/hello-world —env TARGET=Defaulted
Provide a NAME for the service

# AND

kn service create gcr.io/knative/hello-world —env TARGET=Defaulted --force
Service 'hello-world' created in 'default' namespace.
```

I like 1. But I could live with 2. I am easy that way 🏖

**Comment by rhuss** [comment]
I tend to favour 2. because if you have conventions as user I start to rely on them. So when a service is called after its image by default, then this should happen all the time. I want to be noticed when this cannot be the case so that I can fix it by giving it another name explicitly.

But as mentioned, I don't see having the same image for multiple services in the same namespace as a frequent use case anyway. (For revisions, this is different, but there you already have an automatic suffixing)

**Comment by duglin** [comment]
I like 2 as well, but I would change the error to complain about the service name already being used, since that's the real error, not that they forgot to give us a name. Which is the same error they would get if they did `kn service create duglin/echo echo` twice in a row.

I agree that option 1 is a bit odd since it doesn't result in what the user asked for, a name that matches the image name. It also isn't deterministic since the resulting name would be based on what's in the system and the user would have the scrap the output to know what was just created - so that's not a great UX.

**Comment by maximilien** [comment]
OK fair enough. I can jump on the option 2 bandwagon. Let’s do it.

**Comment by maximilien** [comment]
Updating option 2 with better messages:

```bash
kn service create gcr.io/knative/hello-world hello-world —env TARGET=Defaulted
Service 'hello-world' created in 'default' namespace.

...

kn service create gcr.io/knative/hello-world —env TARGET=Defaulted
Service `hello-world` already exist, please provide a NAME for the service

# AND

kn service create gcr.io/knative/hello-world —env TARGET=Defaulted --force
Service 'hello-world' re-created in 'default' namespace.
```

**Comment by maximilien** [comment]
Seems like we don't have consensus on option 2 fully after 06/11/2019 call. However, we seem to have agreement that 2 is better than 1 and that most of 2 is OK. In particular, we agreed on:

```bash
kn service create gcr.io/knative/hello-world hello-world —env TARGET=Defaulted
Service 'hello-world' created in 'default' namespace.

# AND

kn service create gcr.io/knative/hello-world —env TARGET=Defaulted --force
Service 'hello-world' re-created in 'default' namespace.
```
The only problem with above, if we use it solely, that is requiring both `IMAGE` and `NAME` is which should go first. This is important since if both are required then should `NAME` be first?

If we agree that `NAME` is optional then the question becomes how to deal with:

```bash
# Assume a service with name `hello-world` exists in namespace
kn service create gcr.io/knative/hello-world —env TARGET=Defaulted
Service 'hello-world' created in `default` namespace
```
The issue with above are:

1. what happens when image name is in a format that cannot be covered to service name, e.g, uses `.` in name? Can this happen
2. what name should be used when image has a version tag?
3. what message to give to user when `--force` option is used

Overall some felt that there would be "too much magic" in this implementation.

cc: @cppforlife @navidshaikh @evankanderson @houshengbo hopefully I captured everything? If not please comment.

**Comment by sixolet** [comment]
Every other command, the resource to operate on is the one and only positional. 

I'm still not convinced to deviate from that.

**Comment by duglin** [comment]
Couple of things that sway my thinking:
1 - I like the pattern of: `kn NOUN VERB NAME-OF-NOUN ...` in general. I think this is very natural for people to remember. So, `kn service create foo ...` is more natural to me than `kn service create <an arg> foo`. I think having the name of the object as the first user-defined arg is important as it helps scope what we're operating on. 
2 - I was ok with moving NAME to the end only because the discussion was talking about making it optional. At that point the pattern I like in the previous bullet takes a back-seat to the desire of NAME being optional.  But, as I said initially, I'm not as keen on that. I personally don't see it as a burden to require a NAME on this operation since that is the unique identifier that the user will use going forward - so being explicit and clear about it is probably a good thing
3 - on today's call it wasn't clear to me that there was consensus for name being optional

So, with all of that, I'd like to propose this:
```
kn service create NAME IMAGE [ OPTIONS ]
```
where technically `[ OPTIONS ]` can be any place after the `create`

**Comment by sixolet** [comment]
>  
> 
> So, with all of that, I'd like to propose this:
> 
> kn service create NAME IMAGE [ OPTIONS ]
> where technically [ OPTIONS ] can be any place after the create
> 

Even better if image can be flag OR positional so that you can port update commands directly to create.

**Comment by duglin** [comment]
there's no need for a "flag" since it's required.  flags are for optional things.
I don't think consistency of `--image` is important or relevant since in one case IMAGE is optional and in the other it's required.  IMO it's more confusing to imply it's optional by requiring a flag.

**Comment by rhuss** [comment]
Before diving in, let me say I'm cool with everything. I'd like to have a good default name if one can be easily extracted and there is a clear 1:1 to the origin (i think that's the case for the image name), but can as well live with a mandatory name (i just saw the default behaviour as a 'naming suggestion' which _does not need to be used_)

I have not many new arguments but let me answer to some of the concerns:

> * what happens when image name is in a format that cannot be covered to service name, e.g, uses `.` in name? Can this happen?

I believe the [allowed character set for K8s resources names](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/) and the ["repository" part of an image name](https://github.com/moby/moby/blob/04da4041757370fb6f85510c8977c5a18ddae380/vendor/github.com/docker/distribution/reference/regexp.go#L18) is quite similar with rare exceptions, like '_' (. seems to be allowed for k8s names, too, but needs to be verified). Translating forbidden chars (which I suspect to be used seldom) to a "_" would be acceptable for me. Or break with an error: "Can't use 'my.strange.image.name' as default service name. Please provide a name on your own" would be also ok

> * what name should be used when image has a version tag?

Should not be considered for the image name. Only the repository name part (that after registry/user and before tag).

> * what message to give to user when `--force` option is used?

The same as now, namely that a service of 'name' is overwritten. We could require a name argument for --force, too, to be sure. So a default name would then be *only* used when nothing destructive is going to be performed.

> Every other command, the resource to operate on is the one and only positional.

Technically that would be true for a default name, too, as you don't _operate on_ a resource, but _create it_. I do see a name like an ID which gets generated during the creation of an object and then is used as the identity. For databases, you also often don't provide the id from the outside but getting one generated on behalf of you (except when you are using 'natural' ids where you then have to deal with uniqueness issues).

> Even better if image can be flag OR positional so that you can port update commands directly to create.

I like to have both options (and also to keep a semantic connection between update and create). Also, I don't see any harm in supporting both.

> there's no need for a "flag" since it's required. flags are for optional things.
> I don't think consistency of `--image` is important or relevant since in one case IMAGE is optional and in the other it's required. IMO it's more confusing to imply it's optional by requiring a flag.

Yes, that was my initial point, but then I think convenience (going from update to create) beats consistency, and if we don't _require_ a flag but _support_ that usage, too, that would be a good compromise (and does not put too much of a extra maintenance burden).

------------

To summarize my POV:

* I would love to have a default name as I think its crystal clear how it would be calculated, but do understand the possible reservations against it.
* I can live with the image name as argument (prefered) or given as a 'flag' (and wouldn't call it 'option' anymore).

As said, I'm cool with anything but think we should touch down on this issue soonish.

**Comment by maximilien** [comment]
Agree @rhuss your summary works for me too. 

I'd say unless you see any descending views might be good to just do it.

**Comment by duglin** [comment]
I think "soonish" has passed :-)

**Comment by rhuss** [comment]
> I think "soonish" has passed :-)

:-P yes, sometimes time scales are stretching a bit

Not sure how valid this is, but if we want to change this still we have to be perfect backwards compatible. As we are about _relaxing_ constraints, I think this could be still possible (i.e. allowing to omit the service name and just require a `--image` option and the service name would be calculated from the image name, if possible. If there is a conflict, then an error is returned).

**Comment by duglin** [comment]
I think 
```
kn service create NAME IMAGE [ OPTIONS ]
```
if backwards compatible IF we also support:
```
kn service create NAME --image IMAGE [ OPTIONS ]
```
I think the arg parser would need to be told that the pattern is really `kn service create NAME [IMAGE] [ OPTIONS ]` and then we'd need to add code to verify that either the optional positional arg IMAGE or the --image flag is there, but I think that's do-able.

**Comment by rhuss** [comment]
This could work if we don't have options with optional values because then its not clear if the argument is the options' value or an extra argument. As the user has still both to specify (and `--image` can be shortened to `-i`), I'm not sure that saving 3 key strokes is worth the effort (implementing, testing, documenting).

On the other I still kind of like the idea to allow just the image and the service name gets auto-created then. E.g. when the NAME argument is detected to be an invalid service name but a valid container image name, the last part of the image name can be used as the service name. This could be implemented backwards compatible as well (e.g. instead of throwing an error on an invalid service name, try to use it as an image name *if* it fits the relative specific pattern of container image names)

**Comment by duglin** [comment]
What would the syntax of the create look like if NAME was optional?  Would it be:
```
kn service create IMAGE [NAME] [OPTIONS]
```
That could work, but it wouldn't be backwards compatible.  Seems to me, the only way to make it so, would be to do something a little funky and say that if `--image` is present then the first positional arg is the NAME, otherwise if `--image` is missing then the first positional arg is IMAGE.  Do-able, just a bit different and I guess the syntax would be written as:
```
kn service create (IMAGE [NAME] | NAME --image IMAGE) [OPTIONS]
```
or
```
kn service create IMAGE [NAME] [OPTIONS]
kn service create NAME --image IMAGE [OPTIONS]
```
Definitely different. We should probably first decide if NAME really should be optional

I'm not sure about having an arg switch from NAME to IMAGE based on whether it's syntactically allowable as a NAME - especially when the same string could be a valid NAME or IMAGE at the same time

**Comment by rhuss** [comment]
No, my suggestion would be:

```
kn service create IMAGE
kn service create NAME --image IMAGE
```

* If the positional argument is not a valid name but an image specification (e.g. a `docker.io/myuser/myapp`), then consider the positional argument to be an image. No `--image` is required then, and the service name is extracted from the image name (`myapp`). The extracted name would then be printed in the output (as we already do)
* If the positional argument is a valid service name (as specified in the k8s spec), then `--image` is required to point to the name.

So there a never two positional arguments and this approach is perfectly backwards compatible, as currently when you provide an invalid service name you get an error. 

----

Another idea is to allow image names with a scheme like `ko://....` and trigger the corresponding build mechanism beforehand. Much like mink would do (cc: @mattmoor ). For this to work in a spec conformant way, we would introduce a concept of _image name handler_ which can be installed like plugins (possible with a new prefix like in `kninh-ko`  or `kninh-buildpack`) ("kninh" for "kn image name handler")). Such handlers could then be also inlined for special kn distributions, as we can do for plugins. Then when an image with a scheme prefix is given, kn would look up the corresponding handler (via the filename) and call it to allow to perform extra actions and that handler would then return the 'real' name. I know this independent of this issue at hand and would work with both variations. If we think this is kind of a good idea, I would open an extra issue for a proposal.

----

When both mechanisms are in place, then a simplest possible way for building and deploy a Knative service "helloworld" would just be:

```
kn service create ko://github.com/mattmoor/examples/http/cmd/helloworld
```

I don't think that can become much shorter.

**Comment by mattmoor** [comment]
Haven't grokked the whole issue, but you can already do:
```
kn service create foo --image=$(ko publish ./cmd/blah)
```

I have some changes brewing in mink that would generalize the current ko/dockerfile/buildpack story to support arbitrary tasks / pipelines (later even custom tasks); the idea is `{resource}://{task-name}/{path}`.  I demoed the early concept at HackyHour last week, but I'd be happy to give it again.  At head, you can try `mink run task TASK-NAME -- --help` to see the rough analog for `mink build[pack]` for arbitrary tasks, but I need to connect the dots.

Ultimately, I need to put this in front of the `tkn` folks, but I'm hoping to arrive at a model that doesn't bake ~anything in out of the box before I do that, but I am sort of finding my own thoughts first.

**Comment by rhuss** [comment]
> kn service create foo --image=$(ko publish ./cmd/blah)

My suggestion above is more or less the same, except that it is wrapped in kn itself (via a 'image plugin' mechanism) and could possibly work on Windows as well without the help of a Unix shell. 

> I have some changes brewing in mink that would generalize the current ko/dockerfile/buildpack story to support arbitrary tasks / pipelines (later even custom tasks);

Does this mean that mink gets a hard dependency on Tekton ? I liked the idea to _also_ use Tekton but has a simple, one-shot build alternative, too. 

Really love the idea of a "build URL scheme" when specifying the image name, generalizing the ko concept and might also work with local files. Hence the idea to allow an 'image handler' to be plugged into kn, with kn being completely agnostic how the image is eventually built.

The requirements for such an 'image handler ' would be:

* Create an image out of a given URL (that e.g. contains the path to remote git repo or anything else needed for the build)
* Push to a registry
* Return the image name with digest

That's all that is needed for kn, to create and update services.

**Comment by csantanapr** [comment]
>with kn being completely agnostic how the image is eventually built

I like this goal of having a working directory, and `kn` can build the image locally or build the image remotely on cluster via tekton like mink
